### PR TITLE
DC-465: cloud plugin deactivate api

### PR DESCRIPTION
**Pull request by freddy61025**
https://opennms.atlassian.net/browse/DC-465

### CODE REVIEW

**Review by nmsrreabel — APPROVED** [review]
this looks good to me

File: plugin-common/src/main/java/org/opennms/plugins/cloud/config/ConfigurationManager.java [hunk 1]
```diff
@@ -288,6 +291,22 @@ public void initGrpcServices(final GrpcConnectionConfig config) {
         }
     }
 
+    private void destroyGrpcServices() {
+        for (GrpcService service : grpcServices) {
+            try {
+                if (service instanceof TsaasStorage) {
```
Comment (patrick-schweizer) [inline plugin-common/src/main/java/org/opennms/plugins/cloud/config/ConfigurationManager.java:297]: why not add _destroy()_ method to GrpcService? That would avoid casting...

File: plugin-common/src/test/java/org/opennms/plugins/cloud/config/ConfigurationManagerTest.java [hunk 2]
```diff
@@ -200,4 +204,31 @@ public void shouldCallConfigureIfAuthenticatedOrConfigured() {
         assertEquals(CONFIGURED, cm.getStatus());
         verify(grpc, times(1)).initGrpc(any()); // this is done in configure()
     }
+
+    @Test
+    public void shouldAbleToDeactive() {
```
Comment (patrick-schweizer) [inline plugin-common/src/test/java/org/opennms/plugins/cloud/config/ConfigurationManagerTest.java:209]: typo: shouldAbleToDeactive()

File: plugin-common/src/test/java/org/opennms/plugins/cloud/config/ConfigurationManagerTest.java [hunk 3]
```diff
@@ -200,4 +204,31 @@ public void shouldCallConfigureIfAuthenticatedOrConfigured() {
...
+        cm.initConfiguration("something");
+        cm.configure();
+        // should be normal
+        cm.checkConnection();
+
+        cm.deactivateKeyConfiguration();
+
+        assertEquals(DEACTIVATED, cm.getStatus());
+        Exception statusException = null;
+        try {
```
Comment (patrick-schweizer) [inline plugin-common/src/test/java/org/opennms/plugins/cloud/config/ConfigurationManagerTest.java:225]: can be replaced with assertThrows(()-> cm.checkConnection(), StatusRuntimeException.class)

File: plugin-core/src/main/java/org/opennms/plugins/cloud/rest/CloudConfigRestServiceImpl.java [hunk 4]
```diff
@@ -60,18 +65,17 @@ public Response putActivationKey(final String keyJson) {
 
     @Override
     public Response putDeactivateKey(final String keyJson) {
```
Comment (patrick-schweizer) [inline plugin-core/src/main/java/org/opennms/plugins/cloud/rest/CloudConfigRestServiceImpl.java:67]: what is keyJson? I don't think we need it?

File: plugin-core/src/main/java/org/opennms/plugins/cloud/rest/CloudConfigRestServiceImpl.java [hunk 5]
```diff
@@ -60,18 +65,17 @@ public Response putActivationKey(final String keyJson) {
...
-            String key = extractKey(keyJson);
-            this.cm.initConfiguration(key);
-            this.cm.configure();
-            this.cm.deactivateKeyConfiguration(key);
-            
-        } catch (Exception e) {
-            return Response
-                    .status(Response.Status.INTERNAL_SERVER_ERROR)
-                    .entity(exceptionToJson(e))
+        if (this.cm.getStatus() == AUTHENTCATED || this.cm.getStatus() == CONFIGURED) {
```
Comment (patrick-schweizer) [inline plugin-core/src/main/java/org/opennms/plugins/cloud/rest/CloudConfigRestServiceImpl.java:68]: I wonder what happens if status is FAILED.
what should happen if we deactivate plugin? I guess the status should still be DEACTIVATED?

**Review by patrick-schweizer — COMMENTED** [review]
Some comments...

**Pull request merged**